cnxepub/models.py

```diff
@@ -467,6 +486,58 @@ def references(self):
             return []
         return self._references
 
+    def fix_html_ids(self, mapping):
+        """Fix links to other pages after auto generated ids.
+        """
+        def replace(match):
+            page_id, fragment = match.groups()
+            new_id = mapping.get(page_id, {}).get(fragment)
+            if new_id:
+                url = match.string.split('#')[0]
+                return '{}#{}'.format(url, new_id)
+
+        for a in self._xml.xpath('//a[@href]'):
+            href = a.attrib['href']
+            a.attrib['href'] = re.sub(
+                '/contents/([^/@]+)/?#(.+)', replace, href)
+
+    def generate_html_ids(self):
+        """Generate unique ids for html elements in the content so that it's
+        possible to link to them.
+        """
+        elements = [
+            'p', 'dl', 'dt', 'dd', 'table', 'div', 'section', 'figure',
+            'blockquote', 'q', 'code', 'pre', 'object', 'img', 'audio',
+            'video',
+            ]
+        elements_xpath = '|'.join(['//{}'.format(elem) for elem in elements])
+
+        data_types = [
+            'equation', 'list', 'exercise', 'rule', 'example', 'note',
+            'footnote-number', 'footnote-ref', 'problem', 'solution', 'media',
+            'proof', 'statement', 'commentary'
+            ]
+        data_types_xpath = '|'.join(['//*[@data-type="{}"]'.format(data_type)
+                                     for data_type in data_types])
+
+        xpath = '|'.join([elements_xpath, data_types_xpath])
+
+        mapping = {}  # old id -> new id
+
+        for i, node in enumerate(self._xml.xpath(xpath)):
+            old_id = node.attrib.get('id')
+            new_id = 'id-{}-{}'.format(self.id, i)
+            node.attrib['id'] = new_id
+            if old_id:
+                mapping[old_id] = new_id
```

Comment:
At the moment, this (id generation) code is only invoked when running the single html script.